File: java/src/main/java/sk/epostak/sdk/resources/InboundResource.java
package sk.epostak.sdk.resources;

import com.google.gson.reflect.TypeToken;
import sk.epostak.sdk.HttpClient;
import sk.epostak.sdk.models.CursorPage;
import sk.epostak.sdk.models.InboundAckParams;
import sk.epostak.sdk.models.InboundDocument;
import sk.epostak.sdk.models.InboundListParams;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Pull API — inbound (received) documents.
 * <p>
 * Provides cursor-paginated listing, single document retrieval, raw UBL download,
 * and idempotent acknowledgement of inbound Peppol documents.
 * <p>
 * Requires {@code requireApiEligiblePlan} (api-enterprise or integrator-managed)
 * and scope {@code documents:read} for read operations, {@code documents:write}
 * for {@link #ack(String, InboundAckParams)}.
 * <p>
 * Access via {@code client.inbound()}.
 *
 * <pre>{@code
 * CursorPage<InboundDocument> page = client.inbound().list(
 *     InboundListParams.builder()
 *         .since("2026-05-01T00:00:00Z")
 *         .limit(100)
 *         .build());
 *
 * for (InboundDocument doc : page.items()) {
 *     String xml = client.inbound().getUbl(doc.getId());
 *     processInvoice(xml);
 *     client.inbound().ack(doc.getId(), InboundAckParams.empty());
 * }
 * }</pre>
 */
public final class InboundResource {

    private static final TypeToken<CursorPage<InboundDocument>> PAGE_TYPE =
            new TypeToken<CursorPage<InboundDocument>>() {};

    private final HttpClient http;

    /**
     * Creates a new inbound resource.
     *
     * @param http the HTTP client used for API communication
     */
    public InboundResource(HttpClient http) {
        this.http = http;
    }

    /**
     * List inbound documents with cursor pagination. Newest first.
     *
     * @param params optional filters; pass {@link InboundListParams#empty()} for defaults
     * @return one page of inbound documents with next-cursor
     * @throws sk.epostak.sdk.EPostakException if the request fails
     */
    public CursorPage<InboundDocument> list(InboundListParams params) {
        Map<String, Object> qp = new LinkedHashMap<>();
        if (params != null) {
            qp.put("since", params.since());
            qp.put("limit", params.limit());
            qp.put("kind", params.kind());
            qp.put("sender", params.sender());
            qp.put("cursor", params.cursor());
        }
        String path = "/inbound/documents" + HttpClient.buildQuery(qp);
        return http.getTyped(path, PAGE_TYPE);
    }

    /**
     * List inbound documents with default parameters.
     *
     * @return first page of inbound documents (server default limit, no filters)
     * @throws sk.epostak.sdk.EPostakException if the request fails
     */
    public CursorPage<InboundDocument> list() {
        return list(InboundListParams.empty());
    }

    /**
     * Retrieve a single inbound document by ID.
     *
     * @param id the inbound document UUID
     * @return the inbound document detail
     * @throws sk.epostak.sdk.EPostakException if the document is not found or the request fails
     */
    public InboundDocument get(String id) {
        return http.get("/inbound/documents/" + HttpClient.encode(id), InboundDocument.class);
    }

    /**
     * Download the raw UBL 2.1 XML for an inbound document.
     * <p>
     * Returns {@code 404} when the document has no stored UBL (legacy rows).
     *
     * @param id the inbound document UUID
     * @return the raw UBL XML string
     * @throws sk.epostak.sdk.EPostakException if the document or its UBL is not found
     */
    public String getUbl(String id) {
        return http.getString("/inbound/documents/" + HttpClient.encode(id) + "/ubl");
    }

    /**
     * Acknowledge an inbound document (mark it as processed).
     * <p>
     * This operation is idempotent. Re-acknowledging an already-acknowledged document
     * overwrites the {@code clientAckedAt} timestamp with the latest call time
     * (latest-ack-wins semantics).
     * <p>
     * Requires scope {@code documents:write}.
     *
     * @param id     the inbound document UUID
     * @param params optional params including a {@code clientReference} (max 256 chars)
     * @return the updated inbound document with {@code clientAckedAt} set
     * @throws sk.epostak.sdk.EPostakException if the document is not found or the request fails
     */
    public InboundDocument ack(String id, InboundAckParams params) {
        Map<String, Object> body = new LinkedHashMap<>();
        if (params != null && params.clientReference() != null) {
            body.put("client_reference", params.clientReference());
        }
        return http.post(
                "/inbound/documents/" + HttpClient.encode(id) + "/ack",
                body.isEmpty() ? null : body,
                InboundDocument.class
        );
    }

    /**
     * Acknowledge an inbound document without providing a client reference.
     *
     * @param id the inbound document UUID
     * @return the updated inbound document with {@code clientAckedAt} set
     * @throws sk.epostak.sdk.EPostakException if the document is not found or the request fails
     */
    public InboundDocument ack(String id) {
        return ack(id, InboundAckParams.empty());
    }
}

File: dotnet/src/EPostak/Resources/InboundResource.cs
using EPostak.Models;

namespace EPostak.Resources;

/// <summary>
/// Pull API for received (inbound) documents.
/// Use this resource to list, retrieve, download UBL XML, and acknowledge
/// inbound Peppol documents.
/// Requires an API key with the <c>documents:read</c> scope
/// and an <c>api-enterprise</c> or <c>integrator-managed</c> plan.
/// </summary>
public sealed class InboundResource
{
    private readonly HttpRequestor _http;

    internal InboundResource(HttpRequestor http) => _http = http;

    /// <summary>
    /// List inbound documents using cursor-based pagination.
    /// Returns up to <c>Limit</c> (default 100, max 500) documents, newest first.
    /// Advance through pages by passing the returned <c>NextCursor</c> as
    /// <see cref="InboundListParams.Cursor"/> in the next call.
    /// </summary>
    /// <param name="params">Optional filters: cursor, limit, kind, sender, since.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>A page of inbound documents with cursor and <c>HasMore</c> flag.</returns>
    /// <example>
    /// <code>
    /// string? cursor = null;
    /// do
    /// {
    ///     var page = await client.Inbound.ListAsync(new InboundListParams { Cursor = cursor, Limit = 100 });
    ///     foreach (var doc in page.Documents)
    ///         Console.WriteLine($"{doc.Id}: {doc.Kind} from {doc.SenderPeppolId}");
    ///     cursor = page.NextCursor;
    /// } while (page.HasMore);
    /// </code>
    /// </example>
    public Task<InboundListResponse> ListAsync(InboundListParams? @params = null, CancellationToken ct = default)
    {
        var qs = HttpRequestor.BuildQuery(
            ("cursor", @params?.Cursor),
            ("limit", @params?.Limit?.ToString()),
            ("kind", @params?.Kind),
            ("sender", @params?.Sender),
            ("since", @params?.Since));
        return _http.RequestAsync<InboundListResponse>(HttpMethod.Get, $"/inbound/documents{qs}", ct);
    }

    /// <summary>
    /// Get a single inbound document by UUID.
    /// Returns 404 if the document does not belong to the authenticated firm.
    /// </summary>
    /// <param name="id">Inbound document UUID.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The inbound document with all metadata and party details.</returns>
    /// <example>
    /// <code>
    /// var doc = await client.Inbound.GetAsync("doc_uuid");
    /// Console.WriteLine($"Received: {doc.IssueDate}, Amount: {doc.AmountDue} {doc.Currency}");
    /// </code>
    /// </example>
    public Task<InboundDocument> GetAsync(string id, CancellationToken ct = default)
        => _http.RequestAsync<InboundDocument>(HttpMethod.Get, $"/inbound/documents/{Uri.EscapeDataString(id)}", ct);

    /// <summary>
    /// Download the raw UBL 2.1 XML for an inbound document.
    /// Returns the original Peppol BIS 3.0 XML as received on the network.
    /// Returns 404 for legacy rows that pre-date raw XML storage.
    /// </summary>
    /// <param name="id">Inbound document UUID.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>UBL 2.1 XML string (<c>application/xml</c>).</returns>
    /// <example>
    /// <code>
    /// var xml = await client.Inbound.GetUblAsync("doc_uuid");
    /// File.WriteAllText("received_invoice.xml", xml);
    /// </code>
    /// </example>
    public Task<string> GetUblAsync(string id, CancellationToken ct = default)
        => _http.RequestStringAsync(HttpMethod.Get, $"/inbound/documents/{Uri.EscapeDataString(id)}/ubl", ct);

    /// <summary>
    /// Acknowledge an inbound document. Marks the document as processed in the
    /// Pull API (status transitions from <c>received</c> to <c>acked</c>).
    /// This is idempotent — calling it again overwrites <c>ClientAckedAt</c>
    /// and <c>ClientReference</c> with the latest values.
    /// Requires the <c>documents:write</c> scope.
    /// </summary>
    /// <param name="id">Inbound document UUID.</param>
    /// <param name="params">Optional body with <c>ClientReference</c> (max 256 chars).</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The full document with updated <c>ClientAckedAt</c> and <c>ClientReference</c>.</returns>
    /// <example>
    /// <code>
    /// var doc = await client.Inbound.AckAsync("doc_uuid", new InboundAckParams
    /// {
    ///     ClientReference = "our-internal-id-12345"
    /// });
    /// Console.WriteLine($"Acked at: {doc.ClientAckedAt}, ref: {doc.ClientReference}");
    /// </code>
    /// </example>
    public Task<InboundDocument> AckAsync(string id, InboundAckParams? @params = null, CancellationToken ct = default)
        => _http.RequestAsync<InboundDocument>(HttpMethod.Post, $"/inbound/documents/{Uri.EscapeDataString(id)}/ack", @params ?? new InboundAckParams(), ct);
}
